freud/diffraction.pyx

```diff
@@ -0,0 +1,313 @@
+        cdef double complex[:, :] diffraction_fft
+        diffraction_fft = np.fft.fft2(im)
+        diffraction_fft = scipy.ndimage.fourier.fourier_gaussian(
+            diffraction_fft, self.peak_width / self.zoom)
+        diffraction_fft = np.fft.fftshift(diffraction_fft)
+
+        # Compute the squared modulus of the FFT, which is S(q)
+        self._diffraction = np.real(
+            diffraction_fft * np.conjugate(diffraction_fft))
+
+        # Transform the image (scale, shear, zoom) and normalize S(q) by N^2
+        N = len(system.points)
+        self._diffraction = self._transform(
+            self._diffraction, system.box, inv_shear) / (N*N)
+
+        # Compute a cached array of k-vectors that can be rotated and scaled
+        if not self._called_compute:
+            # Create a 1D axis of k-vector magnitudes
+            self._k_values_orig = np.fft.fftshift(np.fft.fftfreq(
+                n=self.output_size,
+                d=1/self.output_size))
+
+            # Create a 3D meshgrid of k-vectors with shape
+            # (output_size, output_size, 3)
+            self._k_vectors_orig = np.asarray(np.meshgrid(
+                self._k_values_orig, self._k_values_orig, [0])).T[0]
+
+        # Compute the rotated and scaled k-values and k-vectors
+        self._k_values = self._k_values_orig / np.max(system.box.to_matrix())
+        self._k_vectors = rowan.rotate(view_orientation, self._k_vectors_orig)
+        self._k_vectors /= np.max(system.box.to_matrix())
+
+        return self
+
+    @_Compute._computed_property
+    def diffraction(self):
+        """:class:`numpy.ndarray`: diffraction pattern."""
+        return self._diffraction
+
+    @_Compute._computed_property
+    def k_values(self):
+        """(:math:`N`, ) :class:`numpy.ndarray`: k-values."""
+        return self._k_values
+
+    @_Compute._computed_property
+    def k_vectors(self):
+        """(:math:`N`, :math:`N`, 3) :class:`numpy.ndarray`: k-vectors."""
+        return self._k_vectors
+
+    def __repr__(self):
+        return ("freud.diffraction.{cls}(grid_size={grid_size}, "
+                "output_size={output_size}, zoom={zoom}, "
+                "peak_width={peak_width})").format(
+                    cls=type(self).__name__,
+                    grid_size=self.grid_size,
+                    output_size=self.output_size,
+                    zoom=self.zoom,
+                    peak_width=self.peak_width)
+
+    def to_image(self, cmap='afmhot', vmin=4e-6, vmax=0.7):
+        """Generates image of diffraction pattern.
+
+        Args:
+            cmap (str):
+                Colormap name to use (Default value = :code:`'afmhot'`).
+            vmin (float):
+                Minimum of the color scale (Default value = 4e-6).
+            vmax (float):
+                Maximum of the color scale (Default value = 0.7).
+
+        Returns:
+            ((output_size, output_size, 4) :class:`numpy.ndarray`):
```

Comment:
Are we sure this returns something with dimensions of `output_size`? The `output_size` variable is never used in this method, and the diffraction pattern doesn't have dimensions of output size.

Reply:
@tommy-waltmann please validate this -- the `_transform` method should return `(output_size, output_size)`. https://github.com/glotzerlab/freud/blob/58ab65f3a20f001dc9da7b02376f7a9162a012d9/freud/diffraction.pyx#L169

Reply:
This is another suggestion I would not have made had had I known the arrays in the constructor were initialized incorrectly.